agasc/supplement/utils.py

```diff
@@ -134,3 +144,191 @@ def save_version(filename, table_name):
                    append=True, overwrite=True)
     last_updated.write(str(filename), format='hdf5', path='last_updated',
                        append=True, overwrite=True)
+
+
+def update_table(filename, table, path, dtype, keys=['agasc_id'], dry_run=False, create=False):
+    """
+    Update a table of the AGASC supplement.
+
+    This overwrites all rows already in the supplement table and appends that rows not yet in the
+    supplement table.
+
+    :param filename:
+    :param table: table.Table
+    :param path: str
+        the path of the table in the supplement ('obs', 'mags')
+    :param dtype: np.dtype
+        the dtype of the table. Used only if the table is not in the supplement already.
+        If it is not given, and the table does not exist in the supplement,
+        an exception will be raised.
+    :param keys: list
+        a list of columns that uniquely identify a row in the table.
+        In case of duplicates, the first appearance is kept.
+    :param dry_run: bool
+    :param create: bool
+        Create a supplement file if it does not exist
+    """
+    filename = Path(filename)
+
+    if len(table) == 0:
+        logger.info('Nothing to update')
+        return
+
+    suppl_table = get_table(filename, path, dtype, create=create)
+    table = Table(table if len(table) else None, dtype=dtype)
+    table = unique(table, keys=keys, keep='first')
+
+    # for now I require no masked elements.
+    if table.mask is not None and np.any(table.mask):
+        masked = [name for name in table.colnames if np.any(table[name].mask)]
+        raise Exception('"{path}" table should have no masked elements, '
+                        'but element in these columns are masked: ',
+                        ', '.join(masked))
+
+    # entries already in supplement
```

Comment:
I guess this works OK here because of the sizes, but in general I'm worried about this approach as being `O(N*M)` in operations and memory. I think an equivalent approach that is `O(N + M)` would be using a dict to make an index e.g (but this is untested of course).
```
index_suppl = {key: i for i, key in enumerate(suppl_table.iterrows(*keys))}

i = []
j = []
k = []
for i_table, key in enumerate(table.iterrows(*keys)):
    try:
        i.append(index_suppl[key])
        j.append(i_table)
    except KeyError:
        k.append(i_table)
```

Reply:
This should not be O(N*M) in terms of memory, because numpy does not actually create the broadcasted arrays. It only reshapes by adding an extra dimension of length 1.

It will be O(N*M) in terms of computation, because it compares all pairs.

In our case, I don't think it matters at all.

Reply:
Let me know if you want me to change this, because I would leave it the way it is.

Reply:
The `intersect` array is N*M, no? 

You can leave this here, but in general for this kind of algorithm which is roughly set intersections you should look for an O(N+M) approach. The first time you put in a couple of 1e6 arrays things will go badly wrong.

Reply:
> The intersect array is N*M, no?

I don't think so. if you do `a[:, None]`, the size of the array in memory does not change, only the shape changes. Then when you call `np.argwhere` it does not return an `N*M` array, so numpy should be able to deal with this without allocating more memory.

Reply:
actually you are right, the intersect array is `N*M`.